### PR TITLE
[Feature]: Support 310P device run qwen2.5/3 dense and qwen2.5vl models

### PR DESCRIPTION
**Pull request by Tflowers-0129**
### What this PR does / why we need it?
Add basic 310p support. Only dense models work with eager mode now.

- vLLM version: v0.13.0
- vLLM main: https://github.com/vllm-project/vllm/commit/2f4e6548efec402b913ffddc8726230d9311948d

### CODE REVIEW

File: vllm_ascend/_310p/attention/attention_mask.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+            return self._get_fp16_mask(max_seq_len)
+        return self._base.get_attn_mask(max_seq_len, dtype)
+
+    def get_splitfuse_attn_mask(self) -> torch.Tensor:
+        return self._get_fp16_mask(2048)
+
+    def get_attention_mask(self, model_config) -> torch.Tensor:
+        if getattr(model_config, "runner_type", None) == "pooling":
+            return self._base.get_attn_mask(2048, torch.bool)
+        return self.get_splitfuse_attn_mask()
```
[the remark this answers is not in the document]
Comment (Tflowers-0129) [reply vllm_ascend/_310p/attention/attention_mask.py]: follow the mian branch

File: vllm_ascend/_310p/attention/attention_mask.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                                   device: torch.device) -> torch.Tensor:
+    tril = torch.ones((max_seq_len, max_seq_len),
+                      dtype=torch.bool,
+                      device=device).tril_()
+    upper = ~tril
+    m = torch.zeros((max_seq_len, max_seq_len),
+                    dtype=torch.float16,
+                    device=device)
+    m.masked_fill_(upper, float("-inf"))
+    return m
```
Comment (csoulndai) [inline vllm_ascend/_310p/attention/attention_mask.py:35]: 多次推理是否会产生碎片显存，是否可以参考这种方式，避免中间张量
m = torch.where(
    torch.tril(torch.ones((max_seq_len, max_seq_len), dtype=torch.bool, device=device)),
    torch.tensor(0.0, dtype=torch.float16, device=device),
    torch.tensor(float("-inf"), dtype=torch.float16, device=device)
)

File: vllm_ascend/_310p/modelrunner_310p.py [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+
+class NPUModelRunner310(NPUModelRunner):
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._acl_format = ACL_FORMAT_FRACTAL_NZ
+
+    def _num_attn_module(self) -> int:
+        return 2 if self.model_config.hf_config.model_type == "longcat_flash" else 1
```
Comment (csoulndai) [inline vllm_ascend/_310p/modelrunner_310p.py]: 魔鬼数字？可以添加注释解释一下吗？

File: vllm_ascend/_310p/worker_310p.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+        torch_npu.npu.set_compile_mode(jit_compile=False)
+
+        from vllm_ascend.worker.worker import init_workspace_manager
+        init_workspace_manager(self.device, num_ubatches=1)
+
+        self.model_runner = NPUModelRunner310(self.vllm_config, self.device)
+
+    def _warm_up_atb(self):
+        logger.info("Skip warm-up atb ops for 310P device")
```
Comment (csoulndai) [inline vllm_ascend/_310p/worker_310p.py:37]: 这里这样写是因为父类写死了报错信息吗

File: vllm_ascend/_310p/attention/attention_v1.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+        if attn_metadata.attn_state == AscendAttentionState.PrefillNoCache:
+            num_tokens = query.shape[0]
+            q = query[:num_tokens]
+            k = key[:num_tokens]
+            v = value[:num_tokens]
+            out = self._forward_prefill_310p_fallback(q, k, v, attn_metadata,
+                                                      output)
+            output[:num_tokens] = out
+
```
Comment (Tflowers-0129) [inline vllm_ascend/_310p/attention/attention_v1.py]: 兜底一下，其他状态raise

File: vllm_ascend/_310p/worker_310p.py [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+
+
+class NPUWorker310(NPUWorker):
+
+    def init_device(self):
+        self.device = self._init_device()
+
+        torch_npu.npu.set_compile_mode(jit_compile=False)
+
+        from vllm_ascend.worker.worker import init_workspace_manager
```
Comment (csoulndai) [inline vllm_ascend/_310p/worker_310p.py]: 这里的引用是不是应该放到文件顶部避免多次查找引用的问题

File: vllm_ascend/_310p/worker_310p.py [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+        init_workspace_manager(self.device, num_ubatches=1)
```
Comment (csoulndai) [inline vllm_ascend/_310p/worker_310p.py:31]: 初始化阶段是否要加一些注释，和日志信息

File: vllm_ascend/_310p/modelrunner_310p.py [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+
+    def _initialize_kv_cache_tensors_310p(
+            self, kv_cache_config: "KVCacheConfig") -> dict[str, Any]:
+        from vllm.v1.kv_cache_interface import FullAttentionSpec
+        from vllm.v1.worker.utils import bind_kv_cache
```
Comment (Tflowers-0129) [inline vllm_ascend/_310p/modelrunner_310p.py]: import 放到顶层

File: vllm_ascend/_310p/attention/attention_mask.py [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    def get_attn_mask(self, max_seq_len: int, dtype: torch.dtype):
+        if dtype == torch.float16:
+            return self._get_fp16_mask(max_seq_len)
+        return self._base.get_attn_mask(max_seq_len, dtype)
+
+    def get_splitfuse_attn_mask(self) -> torch.Tensor:
+        return self._get_fp16_mask(2048)
+
+    def get_attention_mask(self, model_config) -> torch.Tensor:
+        if getattr(model_config, "runner_type", None) == "pooling":
```
Comment (Tflowers-0129) [inline vllm_ascend/_310p/attention/attention_mask.py:92]: 坏味道：夸夸其谈未来

File: vllm_ascend/_310p/attention/attention_mask.py [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+                      device=device).tril_()
+    upper = ~tril
+    m = torch.zeros((max_seq_len, max_seq_len),
+                    dtype=torch.float16,
+                    device=device)
+    m.masked_fill_(upper, float("-inf"))
+    return m
+
+
+class _AttentionMaskBuilder310P:
```
Comment (YangShuai52) [inline vllm_ascend/_310p/attention/attention_mask.py:63]: Can it be designed as a singleton?

File: vllm_ascend/_310p/ops/mm_encoder_attention.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+            ed = st + seg_len
+
+            q_i = q[st:ed].unsqueeze(0)  # [1, S, H, D]
+            k_i = k[st:ed].unsqueeze(0)
+            v_i = v[st:ed].unsqueeze(0)
+
+            qs = int(q_i.shape[1])
+            kvs = int(k_i.shape[1])
+
+            out_i = torch_npu.npu_prompt_flash_attention(
```
Comment (Tflowers-0129) [inline vllm_ascend/_310p/ops/mm_encoder_attention.py:88]: 替换：_npu_flash_attention_unpad

File: vllm_ascend/_310p/ops/activation.py [hunk 12]
```diff
@@ -0,0 +1,15 @@
...
+from vllm_ascend.ops.activation import AscendSiluAndMul as _Base
+
+
+class AscendSiluAndMul310(_Base):
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        torch.ops.vllm.maybe_prefetch_mlp_down_proj(x)
+        h = x.shape[-1] // 2
+        out = (F.silu(x[..., :h].to(torch.float32)) *
+               x[..., h:].to(torch.float32)).to(torch.float16)
```
Comment (Tflowers-0129) [inline vllm_ascend/_310p/ops/activation.py]: 替换atb swiglu融合算子

File: vllm_ascend/_310p/attention/attention_mask.py [hunk 13]
```diff
@@ -0,0 +1,61 @@
+from typing import Any, Callable, Optional
+
+import torch
+
+import vllm_ascend.attention.attention_mask as _base_mask
+
+_BASE_BUILDER: Callable[[torch.device], Any] = _base_mask.AttentionMaskBuilder
+
+
```
Comment (Tflowers-0129) [inline vllm_ascend/_310p/attention/attention_mask.py:29]: add comments : list of support mask type

File: vllm_ascend/_310p/attention/attention_mask.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+def _gen_causal_additive_mask_fp16(max_seq_len: int,
```
Comment (Tflowers-0129) [inline vllm_ascend/_310p/attention/attention_mask.py]: func name should change or import from vllm-ascend.attn_mask

File: vllm_ascend/_310p/modelrunner_310p.py [hunk 15]
```diff
@@ -0,0 +1,107 @@
...
+                if layer_name in self.runner_only_attn_layers:
+                    continue
+
+                tensor_size = kv_cache_sizes[layer_name]
+                assert tensor_size % kv_cache_spec.page_size_bytes == 0
+                num_blocks = tensor_size // kv_cache_spec.page_size_bytes
+                assert num_blocks >= kv_cache_config.num_blocks
+
+                if self.vllm_config.additional_config.get(
+                        "kv_cache_dtype", None) == "int8":
```
Comment (Tflowers-0129) [inline vllm_ascend/_310p/modelrunner_310p.py]: not support this dype:int8

File: vllm_ascend/_310p/ops/activation.py (hunk 12 above)
Comment (Tflowers-0129) [reply vllm_ascend/_310p/ops/activation.py]: cannot find

File: vllm_ascend/_310p/attention/attention_v1.py (hunk 5 above)
Comment (Tflowers-0129) [reply vllm_ascend/_310p/attention/attention_v1.py]: Updated

File: vllm_ascend/_310p/attention/attention_v1.py [hunk 16]
```diff
@@ -0,0 +1,113 @@
...
+        return output
```
[the remark this answers is not in the document]
Comment (Tflowers-0129) [reply vllm_ascend/_310p/attention/attention_v1.py]: Updated

File: vllm_ascend/_310p/ops/mm_encoder_attention.py (hunk 11 above)
Comment (Tflowers-0129) [reply vllm_ascend/_310p/ops/mm_encoder_attention.py:88]: torch_npu._npu_flash_attention_unpad operator has precision issues on the 310P device.

File: vllm_ascend/_310p/attention/attention_mask.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+    def get_splitfuse_attn_mask(self) -> torch.Tensor:
+        return self._get_fp16_mask(2048)
+
+    def get_attention_mask(self, model_config) -> torch.Tensor:
+        if getattr(model_config, "runner_type", None) == "pooling":
+            return self._base.get_attn_mask(2048, torch.bool)
+        return self.get_splitfuse_attn_mask()
+
+
+def AttentionMaskBuilder(device: torch.device) -> _AttentionMaskBuilder310P:
```
Comment (ShiyaNiu) [inline vllm_ascend/_310p/attention/attention_mask.py]: AttentionMaskBuilder，框架是类，310p这里的实现做了实例化；是出于什么考虑，AttentionMaskBuilder的设计能否保持一致

File: vllm_ascend/_310p/attention/attention_mask.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Tflowers-0129) [reply vllm_ascend/_310p/attention/attention_mask.py]: updated, 310p device doesnot need to run in "get_splitfuse_attn_mask".

File: vllm_ascend/_310p/modelrunner_310p.py (hunk 3 above)
Comment (Tflowers-0129) [reply vllm_ascend/_310p/modelrunner_310p.py]: 310p devices donot need the num_attn_module because they not run in "longcat_flash" mode

File: vllm_ascend/_310p/worker_310p.py (hunk 6 above)
Comment (Tflowers-0129) [reply vllm_ascend/_310p/worker_310p.py]: update

File: vllm_ascend/_310p/attention/attention_mask.py (hunk 14 above)
Comment (Tflowers-0129) [reply vllm_ascend/_310p/attention/attention_mask.py]: update

**Review by wangxiyuan — APPROVED** [review]
